src/xdr_ioq.c

```diff
@@ -187,6 +191,188 @@ xdr_ioq_uv_recycle(struct poolq_head *ioqh, struct poolq_entry *have)
+		__func__, count, comment, xd);
+
+	pthread_mutex_lock(&ioqh->qmutex);
+
+	while (1) {
+		/* positive for buffer(s) */
+		have = TAILQ_FIRST(&ioqh->qh);
+		__warnx(TIRPC_DEBUG_FLAG_XDR,
+			"%s() %u %s xioq %p %d ioqh %p %d have %p",
+			__func__, count, comment, xioq, xioq->ioq_uv.uvqh.qcount,
+			ioqh, ioqh->qcount, have);
+		if (have) {
+			TAILQ_REMOVE(&ioqh->qh, have, q);
+
+			/* added directly to the queue.
+			 * this lock is needed for context header queues,
+			 * but is not a burden on uncontested data queues.
+			 */
+			pthread_mutex_lock(&xioq->ioq_uv.uvqh.qmutex);
+			(xioq->ioq_uv.uvqh.qcount)++;
+			__warnx(TIRPC_DEBUG_FLAG_XDR, "ioq_track xdr_ioq_uv_fetch insert "
+				"have %p to q %p", have, xioq);
+			TAILQ_INSERT_TAIL(&xioq->ioq_uv.uvqh.qh, have, q);
+			pthread_mutex_unlock(&xioq->ioq_uv.uvqh.qmutex);
+			ioqh->qcount--;
+			count--;
+			if(count == 0)
+				break;
+		} else {
+			if (xd) {
+				pthread_mutex_unlock(&ioqh->qmutex);
+
+				if (ioqh == &xd->inbufs_data.uvqh) {
+					xdr_rdma_add_inbufs_data(xd);
+				}
+
+				if (ioqh == &xd->outbufs_data.uvqh) {
+					xdr_rdma_add_outbufs_data(xd);
+				}
+
+				if (ioqh == &xd->inbufs_hdr.uvqh) {
+					xdr_rdma_add_inbufs_hdr(xd);
+				}
+
+				if (ioqh == &xd->outbufs_hdr.uvqh) {
+					xdr_rdma_add_outbufs_hdr(xd);
+				}
+
+				if (unlikely(ioqh == &xd->cbqh)) {
+					__warnx(TIRPC_DEBUG_FLAG_EVENT, "cbc buffers exhausetd xd %p "
+						"ioqh %p qcount %d", xd, ioqh, ioqh->qcount);
+					rpc_rdma_allocate_cbc_locked(ioqh);
+				}
+
+				pthread_mutex_lock(&ioqh->qmutex);
+			}
+		}
+	}
+
+	pthread_mutex_unlock(&ioqh->qmutex);
+	return have;
+}
+
+struct poolq_entry *
+xdr_rdma_ioq_uv_fetch_nothing(struct xdr_ioq *xioq, struct poolq_head *ioqh,
+			 char *comment, u_int count, u_int ioq_flags)
+{
+	return NULL;
+}
+
+static inline void
+xdr_rdma_ioq_uv_recycle(struct poolq_head *ioqh, struct poolq_entry *have)
+{
+	pthread_mutex_lock(&ioqh->qmutex);
+
+	__warnx(TIRPC_DEBUG_FLAG_XDR, "%s() ioq_track Reccycle ioqh %p %d have %p",
+		__func__, ioqh, ioqh->qcount, have);
+
+	TAILQ_INSERT_TAIL(&ioqh->qh, have, q);
+	ioqh->qcount++;
+
+	pthread_mutex_unlock(&ioqh->qmutex);
+}
+
+void
+xdr_rdma_ioq_uv_release(struct xdr_ioq_uv *uv)
+{
+	uv->u.uio_references = 1;	/* keeping one */
+	xdr_rdma_ioq_uv_recycle(uv->u.uio_p1, &uv->uvq);
+}
+
+void
+xdr_rdma_ioq_release(struct poolq_head *ioqh, bool xioq_recycle,
+    struct xdr_ioq *xioq)
+{
+	struct poolq_entry *have = TAILQ_FIRST(&ioqh->qh);
+
+	/* release queued buffers */
+	while (have) {
+		struct poolq_entry *next = TAILQ_NEXT(have, q);
+
+		assert(xioq->rdma_ioq);
+
+		__warnx(TIRPC_DEBUG_FLAG_XDR, "xdr_rdma_ioq_release ioqh %p %d "
+		    "xioq %p have %p",
+		    ioqh, ioqh->qcount, xioq, have);
+
+		pthread_mutex_lock(&ioqh->qmutex);
+
+		TAILQ_REMOVE(&ioqh->qh, have, q);
+		(ioqh->qcount)--;
+
+		pthread_mutex_unlock(&ioqh->qmutex);
+
+		xdr_rdma_ioq_uv_release(IOQ_(have));
+		have = next;
+	}
+	assert(ioqh->qcount == 0);
```

Comment:
This one has the same issue as the last one.
`have = TAILQ_FIRST(&ioqh->qh);`
would fix it.

Reply:
done